### ISSUE TITLE
Unexpected behaviour

### ISSUE THREAD

**Issue by noughtmare** [typelead/eta#171]
I wrote some bindings to the `java.util.zip.Adler32` library, but I am experiencing some issues. Here's the code:

```haskell
{-# LANGUAGE MagicHash #-}
module Main where

import Java
import Data.Word (Word8)
import GHC.CString

-- ================================================================================
-- java.util.zip.Adler32
-- ================================================================================

data {-# CLASS "java.util.zip.Adler32" #-} Adler32 = Adler32 (Object# Adler32)
  deriving Class

foreign import java unsafe "@new" newAdler32 :: Java a Adler32
foreign import java unsafe getValue :: Java Adler32 Int64
foreign import java unsafe reset :: Java Adler32 ()
foreign import java unsafe "update" updateInt :: Int -> Java Adler32 () 
foreign import java unsafe "update" updateBytes :: JByteArray -> Java Adler32 () 
foreign import java unsafe "update" updateBytesOffLen :: JByteArray -> Int -> Int -> Java Adler32 () 

-- Version of updateInt that works as expected
-- updateInt only uses the lower 8 bytes of its input
updateWord8 :: Word8 -> Java Adler32 ()
updateWord8 = updateInt . fromIntegral

main :: IO ()
main = java $ do 
  a <- newAdler32
  io.print =<< a <.> getValue
  a <.> updateWord8 10
  io . print =<< a <.> getValue
```

This should print:
```
1
720907
```
But it prints:
```
1
1
```

Have I done something wrong?

**Comment by rahulmutt** [comment]
No you have done nothing wrong. Looks like a subtle bug. I have a suspicion that `a <.> getValue` is getting shared. I took a look at the generated code and it's not executing `getValue()` twice as it should. Will investigate further.

**Comment by rahulmutt** [comment]
```haskell
{-# LANGUAGE MagicHash #-}
module Main where

import Java
import Data.Word (Word8)
import GHC.CString
import System.IO

-- ================================================================================
-- java.util.zip.Adler32
-- ================================================================================

data {-# CLASS "java.util.zip.Adler32" #-} Adler32 = Adler32 (Object# Adler32)
  deriving Class

foreign import java unsafe "@new" newAdler32 :: IO Adler32
foreign import java unsafe getValue :: Java Adler32 Int64
foreign import java unsafe reset :: Java Adler32 ()
foreign import java unsafe "update" updateInt :: Int -> Java Adler32 () 
foreign import java unsafe "update" updateBytes :: JByteArray -> Java Adler32 () 
foreign import java unsafe "update" updateBytesOffLen :: JByteArray -> Int -> Int -> Java Adler32 () 

-- Version of updateInt that works as expected
-- updateInt only uses the lower 8 bytes of its input
updateWord8 :: Word8 -> Java Adler32 ()
updateWord8 = updateInt . fromIntegral

main :: IO ()
main = do
  adler <- newAdler32
  javaWith adler $ do
    io.print =<< getValue
    updateWord8 10
    io . print =<< getValue
```
This gives the correct output. I need to add a warning about mutating Java objects in the docs.

**Comment by rahulmutt** [comment]
This bug is strikingly similar to https://ghc.haskell.org/trac/ghc/ticket/5916. To make things safe, we just need to implement the `runRW#` primitive (present in GHC 8) which is a very clean solution.

**Comment by rahulmutt** [comment]
The final fix involved:
1) Backporting `runRW#` from GHC 8 which was more involved than expected (had to backport a lot of other features too to make sure it doesn't kill perf), but this only took care of taking out `realWorld#` from the array operations.
2) Adding new primops that generate fresh state tokens & object tokens:
```
freshStateToken# :: a -> State# s
freshObjectToken# :: a -> Object# b -> Object# b
freshNullObjectToken# :: a -> Object# b
```
Where the `a`'s are levity-polymorphic (accepts types of kinds `*` and `#`). Eta currently does not support language-level expression of levity polymorphism which is ok - it keeps the types simple.

Basically these primops generate a hard dependency and prevents the Float Out transformation in the simplifier from destroying the meaning of the original program. The cool part is that all these primops are essentially nops when the code generator gets around to it, so they add no runtime cost.
